### PR TITLE
feat: add complex vouchers

### LINKED ISSUE #59: Complex Vouchers
## 📚 Context

Request for proposal #003 https://governance.cartesi.io/t/rfp-003-beginner-solidity-complex-vouchers/90

Need of a contract with advanced behaviors for vouchers.

## ✔️ Solution

Creation of a Complex Voucher contract with targeted, expirable, future, ordered and atomic vouchers.

### PR DESCRIPTION
**Pull request by guidanoli**
Closes #59

### CODE REVIEW

File: onchain/rollups/test/foundry/dapp/CartesiDApp.t.sol [hunk 1]
```diff
@@ -535,6 +548,212 @@ contract CartesiDAppTest is TestBase {
...
+
+        // try to execute a voucher that cannot be executed by the tx.origin
+        bool success = executeVoucher(voucher, proof);
+
+        assertEq(success, false);
+
+        vm.prank(address(this), txOrigin);
+
+        // now execute a voucher that has the tx.origin address on the allowed addresses list
+        success = executeVoucher(voucher, proof);
```
Comment (pedroargento) [inline onchain/rollups/test/foundry/dapp/CartesiDApp.t.sol:602]: Are we missing the assert `success==true` here?

**Review by pedroargento — APPROVED** [review]

**Comment by guidanoli** [comment]
This feature has been postponed to v2, as it can much appreciate from the output unification effort.

File: onchain/rollups/test/foundry/dapp/CartesiDApp.t.sol [hunk 2]
```diff
@@ -535,6 +548,214 @@ contract CartesiDAppTest is TestBase {
...
+
+        success = executeVoucher(voucher, proof);
+        assertEq(success, false);
+
+        rewind(2 hours);
+
+        success = executeVoucher(voucher, proof);
+        assertEq(success, true);
+
+        assertEq(address(complex).balance, 0);
```
Comment (ZzzzHui) [inline onchain/rollups/test/foundry/dapp/CartesiDApp.t.sol:755]: I'm curious about this line. Should we check balance?

**Comment by guidanoli** [comment]
@ZzzzHui This PR is blocked by the `DELEGATECALL` voucher feature.

**Comment by guidanoli** [comment]
We'll have to rewrite this using the to-be-implemented `DELEGATECALL` feature.

**Pull request closed without merging**